golang/btcspv/types_test.go

```diff
@@ -171,3 +171,60 @@ func (suite *TypesSuite) TestValidateSPVProof() {
 		suite.EqualError(err, proofCase.Error)
 	}
 }
+
+func (suite *TypesSuite) TestNewHash160Digest() {
+	input := DecodeIfHex("0x1b60c31dba9403c74d81af255f0c300bfed5faa3")
+	output := Hash160Digest{0x1b, 0x60, 0xc3, 0x1d, 0xba, 0x94, 0x3, 0xc7, 0x4d, 0x81, 0xaf, 0x25, 0x5f, 0xc, 0x30, 0xb, 0xfe, 0xd5, 0xfa, 0xa3}
+
+	digest, err := NewHash160Digest(input)
+	suite.Nil(err)
+	suite.Equal(digest, output)
+}
+
+func (suite *TypesSuite) TestNewHash256Digest() {
+	input := DecodeIfHex("0x1406e05881e299367766d313e26c05564ec91bf721d31726bd6e46e60689539a")
+	output := Hash256Digest{0x14, 0x06, 0xe0, 0x58, 0x81, 0xe2, 0x99, 0x36, 0x77, 0x66, 0xd3, 0x13, 0xe2, 0x6c, 0x05, 0x56, 0x4e, 0xc9, 0x1b, 0xf7, 0x21, 0xd3, 0x17, 0x26, 0xbd, 0x6e, 0x46, 0xe6, 0x06, 0x89, 0x53, 0x9a}
+
+	digest, err := NewHash256Digest(input)
+	suite.Nil(err)
+	suite.Equal(digest, output)
+
+	input = DecodeIfHex("0x4f8b42c22dd3729b519ba6f68d2da7cc5b2d606d05daed5ad5128cc03e6c6358")
+	output = Hash256Digest{0x4f, 0x8b, 0x42, 0xc2, 0x2d, 0xd3, 0x72, 0x9b, 0x51, 0x9b, 0xa6, 0xf6, 0x8d, 0x2d, 0xa7, 0xcc, 0x5b, 0x2d, 0x60, 0x6d, 0x05, 0xda, 0xed, 0x5a, 0xd5, 0x12, 0x8c, 0xc0, 0x3e, 0x6c, 0x63, 0x58}
+
+	digest, err = NewHash256Digest(input)
+	suite.Nil(err)
+	suite.Equal(digest, output)
+}
+
+func (suite *TypesSuite) TestNewRawHeader() {
+	input := DecodeIfHex("0x7bb2b8f32b9ebf13af2b0a2f9dc03797c7b77ccddcac75d1216389abfa7ab3750000000000ffffffffaa15ec17524f1f7bd47ab7caa4c6652cb95eec4c58902984f9b4bcfee444567d0000000000ffff")
+	output := RawHeader{0x7b, 0xb2, 0xb8, 0xf3, 0x2b, 0x9e, 0xbf, 0x13, 0xaf, 0x2b, 0x0a, 0x2f, 0x9d, 0xc0, 0x37, 0x97, 0xc7, 0xb7, 0x7c, 0xcd, 0xdc, 0xac, 0x75, 0xd1, 0x21, 0x63, 0x89, 0xab, 0xfa, 0x7a, 0xb3, 0x75, 0x00, 0x00, 0x00, 0x00, 0x00, 0xff, 0xff, 0xff, 0xff, 0xaa, 0x15, 0xec, 0x17, 0x52, 0x4f, 0x1f, 0x7b, 0xd4, 0x7a, 0xb7, 0xca, 0xa4, 0xc6, 0x65, 0x2c, 0xb9, 0x5e, 0xec, 0x4c, 0x58, 0x90, 0x29, 0x84, 0xf9, 0xb4, 0xbc, 0xfe, 0xe4, 0x44, 0x56, 0x7d, 0x00, 0x00, 0x00, 0x00, 0x00, 0xff, 0xff}
+
+	header, err := NewRawHeader(input)
+	suite.Nil(err)
+	suite.Equal(header, output)
+}
+
+func (suite *TypesSuite) TestHeaderFromRaw() {
+	validHeader := suite.ValidProof.ConfirmingHeader
+	// TODO: Why is the prevhash the reverse endianness of what it should be?
+	validHeader.PrevHash = Hash256Digest{0xc2, 0x38, 0xb6, 0x1, 0x30, 0x8b, 0x72, 0x97, 0x34, 0x6a, 0xb2, 0xed, 0x59, 0x94, 0x2d, 0x7d, 0x7e, 0xce, 0xa8, 0xd2, 0x3a, 0x10, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}
```

Comment:
For some reason, `PrevHash` is the reverse endianness of what it should be.  Not sure what is causing this or how to reverse endianness of `Hash256Digest`.  For now, I just explicitly assigned it to the correct value.

Reply:
it may be a bug in my implementation. what endianness is it supposed to be, and what endianness is it?

Come to think of it, this is the only hash field that is not included both ways and explicitly labeled. Might be a good idea to fix that :thinking:

Reply:
I believe it's supposed to be LE.

Reply:
This would be little endian, right? `0x00000000000000000001103ad2a8ce7e7d2d9459edb26a3497728b3001b638c2`

Reply:
> `0x00000000000000000001103ad2a8ce7e7d2d9459edb26a3497728b3001b638c2`

this is BE

Reply:
Ok, I think I know  the problem then.  How would I reverse endianness of type `Hash256Digest`.  `ReverseEndianness` takes a byte array and I can't figure out how to convert `Hash256Digest` to a byte array.

Reply:
In JSON, the prev hash is BE, so I need to change that to LE before comparing.

Reply:
check what I did in the `Verify` functions

Also, write a convenience function to do this so I stop doing it inline.

```
func ReverseHash256Endianness(h Hash256Digest) Hash256Digest {}
```

Reply:
Never mind.  Figured it out

Reply:
Sorry, I just saw your last comment.  Yeah, I can make a function to do it.

Reply:
Ok, made the function, a test for the function, and replaced all instances that try to do this inline